### ISSUE TITLE
Subgraph validation and normalization

### ISSUE THREAD

**Issue by Jannis** [graphprotocol/graph-cli#148]
# Validation

- [ ] Validate subgraph manifest using a GraphQL schema
- [ ] Validate the subgraph's GraphQL schema
- [ ] Validate ABIs in the subgraph
- [ ] Validate that events are present in ABIs

# Normalization

- [ ] Normalize subgraph manifest (deterministic field and list item order etc.)
- [ ] Normalize event signatures (e.g. `uint` -> `uint256`, no parameter names, no spaces)
- [ ] Normalize mappings (e.g. format with prettier)

**Comment by leodasvacas** [comment]
The normalization plan is under the assumption that we're going to read a yaml, normalize it, output another yaml and hash that. One thing that would complicate this is supporting subgraph manifests in json format, which would require a step converting the json to yaml. But the other way around of normalizing yaml to json seems simpler, and json also has the advantage of being easier to put in a canonical shape since it doesn't care about whitespace.

The point is that if we'll support both yaml and json manifests, we should consider having json be the canonical format because it's simpler.

**Comment by Jannis** [comment]
Agreed! The above doesn't make any assumptions like that though — only the current PR does. But YAML is what we use right now, so it seems appropriate. Plus the logic there should also work for JSON.

**Comment by leodasvacas** [comment]
@Jannis wanna check off all that got done here?

**Comment by Jannis** [comment]
@leodasvacas Done, I would've forgotten :)
